platform/kobo/koreader.sh

```diff
@@ -611,10 +611,20 @@ if [ ${RETURN_VALUE} -ne ${KO_RC_HALT} ]; then
         # if we were called from advboot then we must reboot to go to the menu
         # NOTE: This is actually achieved by checking if KSM or a KSM-related script is running:
         #       This might lead to false-positives if you use neither KSM nor advboot to launch KOReader *without nickel running*.
-        if ! pgrep -f kbmenu >/dev/null 2>&1; then
+        if ! pkill -0 kbmenu; then
             /sbin/reboot
         fi
     fi
+else
+    if [ "${VIA_NICKEL}" = "false" ]; then
+        if pkill -0 kbmenu; then
+            # If we were started by KSM and requested an exit, attempt to *NOT* exit the script,
+            # so as not to re-enter KSM at all, to make sure the device powers off with our own ScreenSaver displayed.
+            # NOTE: This might not be fool-proof, as a poweroff might take longer than that,
+            #       or we might be interrupted early by signals.
+            sleep 30
```

Comment:
That's an order of magnitude of a margin of error, but *soit* as they say in French. :-P

Reply:
Actually what I said isn't quite right, because after you shut the device down you expect to be able to start it back up again.

sleep 2 isn't quite enough to be 100% reliable; sleep 3 seems to do the trick reliably.

I'd round that up to 5 for safety and double it for additional safety. I'm not really sure where tripling it yet again comes into play. ^_^